### PR TITLE
Provide first working example implementation

### PR DESCRIPTION
**Pull request by sven1103**
Displays confirmation or failure notification after calling the subscription service.

### CODE REVIEW

File: subscription-cancel-app/pom.xml [hunk 1]
```diff
@@ -28,7 +28,12 @@
         <dependency>
             <groupId>life.qbic</groupId>
             <artifactId>portal-utils-lib</artifactId>
-            <version>2.2.1</version>
+            <version>2.3.0</version>
+        </dependency>
+        <dependency>
+            <groupId>life.qbic</groupId>
+            <artifactId>data-model-lib</artifactId>
```
Comment (jenniferboedker) [inline subscription-cancel-app/pom.xml]: log4j probably comes in with dml --> need to explicitly add properties for log4j 2.17.0

File: subscription-cancel-app/src/main/groovy/life/qbic/portal/subscription/view/SubscriptionInterface.groovy [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+import com.vaadin.ui.themes.ValoTheme
+import life.qbic.portal.subscription.view.model.SubscriptionModel
+
+/**
+ * <class short description - 1 Line!>
+ *
+ * <More detailed description - When to use, what it solves, etc.>
+ *
+ * @since <version tag>
+ */
```
Comment (jenniferboedker) [inline subscription-cancel-app/src/main/groovy/life/qbic/portal/subscription/view/SubscriptionInterface.groovy]: Missing description

File: subscription-cancel-app/src/main/groovy/life/qbic/portal/subscription/view/SubscriptionInterface.groovy [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        confirmation.setComponentAlignment(project, Alignment.MIDDLE_CENTER)
+        confirmation.setComponentAlignment(toLabel, Alignment.MIDDLE_CENTER)
+        confirmation.setComponentAlignment(email, Alignment.MIDDLE_CENTER)
+        confirmation.setSizeFull()
+        this.addComponent(confirmation)
+    }
+
+    void createFailureNotificationLayout() {
+        failureNotification = new VerticalLayout()
+        Label failure = new Label(VaadinIcons.CLOSE.getHtml(), ContentMode.HTML)
```
Comment (jenniferboedker) [inline subscription-cancel-app/src/main/groovy/life/qbic/portal/subscription/view/SubscriptionInterface.groovy]: From the view of UX design I would also choose an icon that has a circle around it

File: subscription-cancel-app/src/main/groovy/life/qbic/portal/subscription/view/SubscriptionInterface.groovy [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        confirmation.addComponent(email)
+        confirmation.setComponentAlignment(affirmative, Alignment.MIDDLE_CENTER)
+        confirmation.setComponentAlignment(mainInfo, Alignment.MIDDLE_CENTER)
+        confirmation.setComponentAlignment(label, Alignment.MIDDLE_CENTER)
+        confirmation.setComponentAlignment(project, Alignment.MIDDLE_CENTER)
+        confirmation.setComponentAlignment(toLabel, Alignment.MIDDLE_CENTER)
+        confirmation.setComponentAlignment(email, Alignment.MIDDLE_CENTER)
+        confirmation.setSizeFull()
+        this.addComponent(confirmation)
+    }
```
Comment (jenniferboedker) [inline subscription-cancel-app/src/main/groovy/life/qbic/portal/subscription/view/SubscriptionInterface.groovy]: Seems very hard to read to me. I would expect to have a general screen for both states. The content should be roughly the same and also the design. You could have something like mainLayout() and then set the titel content, color and icon and texts in the specific methods.

File: subscription-cancel-app/src/main/groovy/life/qbic/portal/subscription/view/SubscriptionInterface.groovy [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+        failure.addStyleName("failure-red")
+        failure.setSizeUndefined()
+        Label mainInfo = new Label("Cancellation failed!")
+        mainInfo.addStyleName(ValoTheme.LABEL_H1)
+        mainInfo.addStyleName("failure-red")
+        mainInfo.setSizeUndefined()
+        Label label = new Label("You did not cancel you subscription.")
+        label.setStyleName(ValoTheme.LABEL_SMALL)
+        label.setSizeUndefined()
+        Label label2 = new Label("Please write an email to <a href='mailto:support@qbic.zendesk.com'>support@qbic.zendesk.com</a> for help.", ContentMode.HTML)
```
Comment (jenniferboedker) [inline subscription-cancel-app/src/main/groovy/life/qbic/portal/subscription/view/SubscriptionInterface.groovy]: To use less words:

```suggestion
        Label label2 = new Label("Please contact <a href='mailto:support@qbic.zendesk.com'>support@qbic.zendesk.com</a> for help.", ContentMode.HTML)
```

File: subscription-cancel-app/src/main/groovy/life/qbic/portal/subscription/view/SubscriptionInterface.groovy [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+        failureNotification = new VerticalLayout()
+        Label failure = new Label(VaadinIcons.CLOSE.getHtml(), ContentMode.HTML)
+        failure.addStyleName(ValoTheme.LABEL_H1)
+        failure.addStyleName("failure-red")
+        failure.setSizeUndefined()
+        Label mainInfo = new Label("Cancellation failed!")
+        mainInfo.addStyleName(ValoTheme.LABEL_H1)
+        mainInfo.addStyleName("failure-red")
+        mainInfo.setSizeUndefined()
+        Label label = new Label("You did not cancel you subscription.")
```
Comment (jenniferboedker) [inline subscription-cancel-app/src/main/groovy/life/qbic/portal/subscription/view/SubscriptionInterface.groovy]:
```suggestion
        Label label = new Label("Your subscription was not canceled.")
```

File: subscription-cancel-app/src/main/groovy/life/qbic/portal/subscription/view/SubscriptionInterface.groovy (hunk 5 above)
Comment (jenniferboedker) [reply subscription-cancel-app/src/main/groovy/life/qbic/portal/subscription/view/SubscriptionInterface.groovy]: NIT: Variable names like label and label2 are not very meaningful

File: subscription-cancel-app/src/main/groovy/life/qbic/portal/subscription/view/SubscriptionPresenter.groovy [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+import life.qbic.portal.subscription.view.model.SubscriptionModel
+
+
+/**
+ * <class short description - 1 Line!>
+ *
+ * <More detailed description - When to use, what it solves, etc.>
+ *
+ * @since <version tag>
+ */
```
Comment (jenniferboedker) [inline subscription-cancel-app/src/main/groovy/life/qbic/portal/subscription/view/SubscriptionPresenter.groovy]: missing description

File: subscription-cancel-app/src/main/resources/developer.properties [hunk 8]
```diff
@@ -57,3 +57,4 @@ access.unauthenticated.sampleproperty = true
 
 # path of the yaml file containing statistics for statistics-portlet
 statistics.file.path = ...
+services.subscription.url = http://service-registry-test.am10.uni-tuebingen.de:8080/services/notifications
```
Comment (jenniferboedker) [inline subscription-cancel-app/src/main/resources/developer.properties]: I think this file should be added to gitignore

File: subscription-cancel-domain/src/main/groovy/life/qbic/business/subscription/api/SubscriptionService.groovy [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+import life.qbic.business.subscription.CancellationConfirmation
+import life.qbic.business.subscription.exceptions.SubscriptionServiceException
+
+/**
+ * <interface short description - 1 Line!>
+ *
+ * <More detailed description - When to use, what it solves, etc.>
+ *
+ * @since <version tag>
+ */
```
Comment (jenniferboedker) [inline subscription-cancel-domain/src/main/groovy/life/qbic/business/subscription/api/SubscriptionService.groovy:12]: missing description

File: subscription-cancel-domain/src/main/groovy/life/qbic/business/subscription/exceptions/SubscriptionServiceException.groovy [hunk 10]
```diff
@@ -0,0 +1,19 @@
+package life.qbic.business.subscription.exceptions
+
+/**
+ * <class short description - 1 Line!>
+ *
+ * <More detailed description - When to use, what it solves, etc.>
+ *
+ * @since <version tag>
+ */
```
Comment (jenniferboedker) [inline subscription-cancel-domain/src/main/groovy/life/qbic/business/subscription/exceptions/SubscriptionServiceException.groovy:9]: missing description

**Review by jenniferboedker — COMMENTED** [review]
Looks good in general. How can I test this portlet?

Furthermore, I added some comments.

**Comment by sven1103** [comment]
Thank you @jenniferboedker for your detailed review. I have to admit, that this is more a PoC implementation, and I dont get a price for readability.
I will refactor that tomorrow.

File: subscription-cancel-app/src/main/groovy/life/qbic/portal/subscription/persistence/Subscriptions.groovy [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+            log.error("Subscription cancelling failed.")
+            log.error(response.getStatusLine())
+            throw new SubscriptionServiceException("Cancellation failed.")
+        }
+    }
+
+    Subscriptions(String serviceUrlBase) {
+        this.serviceUrlBase = Objects.requireNonNull(serviceUrlBase)
+        this.serviceEndPoint = serviceUrlBase + CANCEL_ENDPOINT
+        log.info(serviceEndPoint)
```
Comment (sven1103) [inline subscription-cancel-app/src/main/groovy/life/qbic/portal/subscription/persistence/Subscriptions.groovy]: todo: remove

Comment (jenniferboedker) [reply subscription-cancel-app/src/main/groovy/life/qbic/portal/subscription/persistence/Subscriptions.groovy]: Still a todo?

File: subscription-cancel-app/src/main/groovy/life/qbic/portal/subscription/view/model/ConfirmationDisplay.groovy [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+        components = new ArrayList<>()
+        buildDisplay()
+    }
+
+    private void buildDisplay() {
+        Label affirmative = new Label(VaadinIcons.CHECK_CIRCLE_O.getHtml(), ContentMode.HTML)
+        affirmative.addStyleName(ValoTheme.LABEL_H1)
+        affirmative.addStyleName(ValoTheme.LABEL_COLORED)
+
+        Label topic = new Label("Subscription cancelled.")
```
Comment (jenniferboedker) [inline subscription-cancel-app/src/main/groovy/life/qbic/portal/subscription/view/model/ConfirmationDisplay.groovy]: NIT: to make it more similar to the cancellation 
```suggestion
        Label topic = new Label("Subscription cancelled!")
```

File: subscription-cancel-app/src/main/groovy/life/qbic/portal/subscription/view/model/ConfirmationDisplay.groovy [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+        topic.addStyleName(ValoTheme.LABEL_COLORED)
+
+        Label info = new Label("We stop sending updates for")
+        info.setStyleName(ValoTheme.LABEL_SMALL)
+
+        project = new Label("PROJECT PLACEHOLDER")
+        project.addStyleName(ValoTheme.LABEL_H3)
+
+        Label toLabel = new Label("to")
+        toLabel.setStyleName(ValoTheme.LABEL_SMALL)
```
Comment (jenniferboedker) [inline subscription-cancel-app/src/main/groovy/life/qbic/portal/subscription/view/model/ConfirmationDisplay.groovy:52]: NIT: I think the to label does not need a separate line. You could place it into the same line as the email

File: subscription-cancel-app/src/main/groovy/life/qbic/portal/subscription/view/model/ConfirmationDisplay.groovy [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+        email = new Label("EMAIL PLACEHOLDER")
+        email.setStyleName(ValoTheme.LABEL_H3)
+
+        components.addAll([affirmative, topic, info, project, toLabel, email])
+        components.stream()
+                .forEach((Component c) -> {
+                    c.setSizeUndefined()
+                    this.addComponent(c)
+                    this.setComponentAlignment(c, Alignment.MIDDLE_CENTER)
+                })
```
Comment (jenniferboedker) [inline subscription-cancel-app/src/main/groovy/life/qbic/portal/subscription/view/model/ConfirmationDisplay.groovy:63]: neat!

File: subscription-cancel-app/src/main/groovy/life/qbic/portal/subscription/view/SubscriptionInterface.groovy [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+    private void setListeners() {
+        model.addPropertyChangeListener("project", {
+            confirmationDisplay.setProject(it.newValue as String)
+        })
+        model.addPropertyChangeListener("email", {
+            confirmationDisplay.setEmail(it.newValue as String)
+        })
+        model.addPropertyChangeListener("showConfirmation", {
+            failureDisplay.setVisible(!it.newValue as Boolean)
+            confirmationDisplay.setVisible(it.newValue as Boolean)
```
Comment (jenniferboedker) [inline subscription-cancel-app/src/main/groovy/life/qbic/portal/subscription/view/SubscriptionInterface.groovy]: I would rather separate it and use the showFailureNotificationLayout property

**Review by jenniferboedker — APPROVED** [review]
Nice work! 🎉

**Pull request merged**